### PR TITLE
[AArch64] Add disassembly for TRACE=ustubs.

### PR DESCRIPTION
**Pull request by dave-estes**

### CODE REVIEW

File: hphp/util/disasm.cpp [hunk 1]
```diff
@@ -146,6 +148,26 @@ void Disasm::disasm(std::ostream& out, uint8_t* codeStartAddr,
...
+  auto end = reinterpret_cast<vixl::Instruction*>(codeEndAddr);
+
+  while (i <= end) {
+    decoder->Decode(i);
+    out << "  " << disasm->GetOutput() << std::endl;
+    i = i->NextInstruction();
+  }
+
+  delete decoder;
+  delete disasm;
```
Comment (mxw) [inline hphp/util/disasm.cpp]: Why can't you use the stack here?

File: hphp/util/disasm.cpp [hunk 2]
```diff
@@ -73,7 +75,7 @@ static const xed_syntax_enum_t s_xed_syntax =
 void Disasm::disasm(std::ostream& out, uint8_t* codeStartAddr,
                     uint8_t* codeEndAddr) {
 
-#ifdef HAVE_LIBXED
+#if (defined __x86_64__) && (defined HAVE_LIBXED)
```
Comment (mxw) [inline hphp/util/disasm.cpp]: Prefer `defined(__x86_64__) && defined(HAVE_LIBXED)`

Comment (dave-estes) [inline hphp/util/disasm.cpp]: Will do.

File: hphp/util/disasm.cpp (hunk 1 above)
Comment (dave-estes) [inline hphp/util/disasm.cpp]: Will do.

File: hphp/util/disasm.cpp [hunk 3]
```diff
@@ -146,6 +148,23 @@ void Disasm::disasm(std::ostream& out, uint8_t* codeStartAddr,
     frontier += instrLen;
     ip       += instrLen;
   }
+
+#elif (defined __aarch64__)
```
Comment (mxw) [inline hphp/util/disasm.cpp]: `defined(__aarch64__)`

**Comment by mxw** [comment]
The test failures are because `hphp/util/` isn't allowed to depend on `hphp/vixl/`.  `util/` is very carefully curated and kept minimal, so I think the best solution here is to add a `jit/disasm.h` module and switch on `arch()` in there, then replace uses of `Disasm` with that.  (We could rename `util/disasm.*` to `util/disasm-x64.*`, perhaps.)

**Comment by dave-estes** [comment]
What do you think about moving hphp/util/asm.* and hphp/util/disasm.*  to .../jit/ and then using switch arch() for their implementations. I confirmed that the only files including disasm and asm were already in .../jit/?

**Comment by dave-estes** [comment]
That was a good idea, @mxw. In fact, there was already a disassembly printer in .../jit/print.cpp, so I just changed unique-stubs.cpp to use that printer instead. As a bonus, it has support for PPC64 now too.

**Comment by mxw** [comment]
You need to rebase this diff—there is no longer such a file as `mc-generator.cpp`.

**Comment by dave-estes** [comment]
It's rebased now. I ended up pushing with --force. I hope that's the right way to get changes up after a rebase.

**Comment by mxw** [comment]
Seems to have worked.  Tests should be good, so I'll land this once they're all clear.

**Pull request closed without merging**